src/input.rs

```diff
@@ -86,7 +84,7 @@ where
         // need to read from this input. Therefore, we wait to acquire
         // the lock on `map` until we also hold the global query write
         // lock.
-        db.salsa_runtime().with_incremented_revision(|next_revision| {
+        db.salsa_runtime().with_incremented_revision(|guard| {
```

Comment:
This comment seems slightly outdated. Additionally, looks like a subtle thing is going on here, which probably warrants a comment as well.

I would expect that we `mark_durability_as_changed` using the maximum of the durabilities of the old and the new slot. However, we advance only the old durability. I think this is ok:

* if there is no previous entry, no one could have observed it, so it's ok to not advance revisions
* if the previous entry has higher durability, that's equivalent to the "obviously correct" max of durabilites
* if it has lower durability, it is correct to not advance the higher durability, because no query of this durability could have observed the input.

This also makes me think that it would even be sound to not bump the base revision when we set previously non-existent key.

Reply:
> This also makes me think that it would even be sound to not bump the base revision when we set previously non-existent key.

I think this is correct. It wasn't always so, as we used to have `Default` key values -- but now we panic on access to a non-existent key, so that's ok.

Reply:
Also, I want (but haven't) add some way to get the "set of valid keys" on an input, in which case we would need a new revision.

Reply:
>Also, I want (but haven't) add some way to get the "set of valid keys" on an input, in which case we would need a new revision.

If we do that, we'll need to modify the logic to always use max durability, without shortcutrs